references/classification/utils.py

```diff
@@ -400,3 +401,57 @@ def reduce_across_processes(val):
     dist.barrier()
     dist.all_reduce(t)
     return t
+
+
+def set_weight_decay(
+    model: torch.nn.Module,
+    weight_decay: float,
+    norm_weight_decay: Optional[float] = None,
+    bias_weight_decay: Optional[float] = None,
+    custom_keys_weight_decay: Optional[Dict[str, float]] = None,
+):
+    norm_classes = (torch.nn.modules.batchnorm._BatchNorm, torch.nn.LayerNorm, torch.nn.GroupNorm)
+
+    norm_params = []
+    bias_params = []
+    other_params = []
+    custom_params = {}
+    if custom_keys_weight_decay is not None:
+        for key in custom_keys_weight_decay:
+            custom_params[key] = []
+
+    for module in model.modules():
+        if next(module.children(), None):
+            for name, p in module.named_parameters(recurse=False):
+                if not p.requires_grad:
+                    continue
+                is_custom_key = False
+                for key in custom_params:
+                    if key in name:
+                        custom_params[key].append(p)
+                        is_custom_key = True
+                if not is_custom_key:
+                    other_params.append(p)
+        elif isinstance(module, norm_classes):
+            if norm_weight_decay is not None:
+                norm_params.extend(p for p in module.parameters() if p.requires_grad)
+            else:
+                other_params.extend(p for p in module.parameters() if p.requires_grad)
+        else:
+            for name, p in module.named_parameters():
+                if not p.requires_grad:
+                    continue
+                if name == "bias" and (bias_weight_decay is not None):
+                    bias_params.append(p)
+                else:
+                    other_params.append(p)
+
+    param_groups = []
+    if norm_weight_decay is not None:
+        param_groups.append({"params": norm_params, "weight_decay": norm_weight_decay})
+    if bias_weight_decay is not None:
+        param_groups.append({"params": bias_params, "weight_decay": bias_weight_decay})
+    for key in custom_params:
+        param_groups.append({"params": custom_params[key], "weight_decay": custom_keys_weight_decay[key]})
+    param_groups.append({"params": other_params, "weight_decay": weight_decay})
+    return param_groups
```

Comment:
Overall, I think this could be a useful utility which can replace the `split_normalization_params` method in ops. I think we should consider refactoring/simplifying and moving it out on a separate PR. It would be a useful addition in the ops though I would still keep it private until we are certain that the API is generic enough.

Reply:
@datumbox I think that we should improve the setting of `weight decay` to support more. what do you think?

Reply:
I think we agree but to clarify what I meant: I see lots of value in making a more generic util that allows us setting custom weigh decay for biases, norm weights and arbitrary params. I think your implementation is a great starting point for this and that's why I've added some comments to refactor it and potentially move it potentially on a separate PR that is going to be merged first on main. Is your understanding the same?

Reply:
@datumbox yes, I argee that we should open anthor PR to add a more generic  to replace `split_normalization_params`.

Reply:
Agreed. Just remove the method from this PR, move it to another one and incorporate the comments we added already here. We will then review the other PR and merge it first before we merge this one. :)